pkg/provisioner/crdprovisioner.go

```diff
@@ -0,0 +1,265 @@
+	parameters map[string]string,
+	secrets map[string]string,
+	volumeContentSource *v1alpha1.ContentVolumeSource,
+	accessibilityReq *v1alpha1.TopologyRequirement) (*v1alpha1.AzVolumeStatusParams, error) {
+	azV := c.azDiskClient.DiskV1alpha1().AzVolumes(c.namespace)
+	azVolume := &v1alpha1.AzVolume{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: volumeName,
+		},
+		Spec: v1alpha1.AzVolumeSpec{
+			UnderlyingVolume:          volumeName,
+			VolumeCapability:          volumeCapabilities,
+			CapacityRange:             capacityRange,
+			Parameters:                parameters,
+			Secrets:                   secrets,
+			ContentVolumeSource:       volumeContentSource,
+			AccessibilityRequirements: accessibilityReq,
+		},
+	}
+
+	azVolumeCreated, err := azV.Create(ctx, azVolume, metav1.CreateOptions{})
+	if err != nil || azVolumeCreated == nil {
+		klog.Errorf("Failed to create azvolume resource for volume name (%s), error: %v", volumeName, err)
+		return nil, err
+	}
+
+	if azVolumeCreated.Status == nil {
+		return nil, status.Error(codes.Internal, fmt.Sprintf("Unable to fetch status of volume created for volume name (%s)", volumeName))
+	}
+
+	return azVolumeCreated.Status.ResponseObject, nil
+}
+
+func (c *CrdProvisioner) DeleteVolume(ctx context.Context, volumeID string, secrets map[string]string) error {
+	azV := c.azDiskClient.DiskV1alpha1().AzVolumes(c.namespace)
+	err := azV.Delete(ctx, volumeID, metav1.DeleteOptions{})
+	if errors.IsNotFound(err) {
+		klog.Info("Could not find the volume name (%s). Deletion succedded", volumeID)
+		return nil
+	}
+
+	if err != nil {
+		klog.Error("Failed to delete azvolume resource for volume id (%s), error: %v", volumeID, err)
+		return err
+	}
+
+	// Verify if the azVolume is deleted
+	azVolume, _ := azV.Get(ctx, volumeID, metav1.GetOptions{})
+	if azVolume != nil {
+		return status.Error(codes.Internal, fmt.Sprintf("Failed to delete azvolume resource for volume id (%s)", volumeID))
+	}
+
+	return nil
+}
+
+func (c *CrdProvisioner) PublishVolume(
+	ctx context.Context,
+	volumeID string,
+	nodeID string,
+	volumeCapability *v1alpha1.VolumeCapability,
+	readOnly bool,
+	secrets map[string]string,
+	volumeContext map[string]string) (*v1alpha1.PublishedVolumeParams, error) {
+	azVA := c.azDiskClient.DiskV1alpha1().AzVolumeAttachments(c.namespace)
+	azVolumeAttachment := &v1alpha1.AzVolumeAttachment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: fmt.Sprintf("%s-%s", volumeID, nodeID),
+		},
+		Spec: v1alpha1.AzVolumeAttachmentSpec{
+			UnderlyingVolume: volumeID,
+			NodeName:         nodeID,
+			VolumeContext:    volumeContext,
+			RequestedRole:    v1alpha1.PrimaryRole,
+		},
+	}
+
+	attachmentCreated, err := azVA.Create(ctx, azVolumeAttachment, metav1.CreateOptions{})
+	if err != nil {
+		klog.Error("Error creating azvolume attachment for volume id (%s) to node id (%s) error : %v", volumeID, nodeID, err)
+		return nil, err
+	}
+
+	if attachmentCreated.Status == nil {
```

Comment:
It will take a bit of time for the controller to initialize Status for the `AzVolumeAttachment` as Status field is initialized once the volume is successfully published via cloud provisioner.

So, some sort of wait would be necessary to check if the publish volume operation has succeeded or not.

Reply:
As per the current changes we update the status of the CRIs only when the operation succeeds,right? . If it fails, we are requeuing the request to be looked at later. So we might not have a definite time until which we should wait for the status to change.

Reply:
I have added a PollImmediate with pre-defined time parameters as of now